Wrappers/Python/test/test_CIL_CVXPy.py

```diff
@@ -0,0 +1,148 @@
+# -*- coding: utf-8 -*-
+#   This work is part of the Core Imaging Library (CIL) developed by CCPi 
+#   (Collaborative Computational Project in Tomographic Imaging), with 
+#   substantial contributions by UKRI-STFC and University of Manchester.
+
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+
+#   http://www.apache.org/licenses/LICENSE-2.0
+
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+from cil.optimisation.functions import L2NormSquared
+from cil.optimisation.functions import TotalVariation
+
+from cil.utilities import dataexample
+from cvxpy import *
+
+import numpy as np
+import scipy.sparse as sp
+
+import unittest
+
+class Test_CIL_vs_CVXPy(unittest.TestCase):
+
+    def setUp(self):
+
+        # default test data
+        self.data = dataexample.CAMERA.get(size=(32, 32))
+
+    
+    # Create gradient operator as sparse matrix that will be used in CVXpy
+    def sparse_gradient_matrix(self, shape, direction='forward', order=1, boundaries='Neumann', **kwargs):
+        
+        len_shape = len(shape)    
+        allMat = dict.fromkeys(range(len_shape))
+        discretization = kwargs.get('discretization',[1.0]*len_shape)
+
+        if order == 1:
+
+            for i in range(0,len_shape):
+
+                if direction == 'forward':
+
+                    mat = 1/discretization[i] * sp.spdiags(np.vstack([-np.ones((1,shape[i])),np.ones((1,shape[i]))]), [0,1], shape[i], shape[i], format = 'lil')
+
+                    if boundaries == 'Neumann':
+                        mat[-1,:] = 0
+                    elif boundaries == 'Periodic':
+                        mat[-1,0] = 1
+
+                elif direction == 'backward':
+
+                    mat = 1/discretization[i] * sp.spdiags(np.vstack([-np.ones((1,shape[i])),np.ones((1,shape[i]))]), [-1,0], shape[i], shape[i], format = 'lil')
+
+                    if boundaries == 'Neumann':
+                        mat[:,-1] = 0
+                    elif boundaries == 'Periodic':
+                        mat[0,-1] = -1
+
+                tmpGrad = mat if i == 0 else sp.eye(shape[0])
+
+                for j in range(1, len_shape):
+
+                    tmpGrad = sp.kron(mat, tmpGrad ) if j == i else sp.kron(sp.eye(shape[j]), tmpGrad )
+
+                allMat[i] = tmpGrad
+
+        else:
+            raise NotImplementedError    
+
+        return allMat        
+
+    def tv_cvxpy_regulariser(self, u, isotropic=True, direction = "forward", boundaries = "Neumann"):
```

Comment:
as the unit tests don't change `direction` and `boundaries` is it worth having this function at all? the 3 lines (with the correct isotropic code) can be used to the test call directly:

``` python
G = self.sparse_gradient_matrix(u.shape)   
DX, DY = G[1], G[0]
return sum(norm(vstack([DX @ vec(u), DY @ vec(u)]), 2, axis = 0))
```

Reply:
To create the Totalvariation function for cvxpy, we need the Sparse gradient matrix, which creates two sparse matrices for DX, DY. This is important not only for TV, but for TGV, dTV. Actually in this script, I will include all the tests comparing CIL and CVXpy and for the above regularisers we need the sparse gradient matrix. This is already developed in CIL but never userd, see (SparseFiniteDifferenceOperator)[https://github.com/TomographicImaging/CIL/blob/master/Wrappers/Python/cil/optimisation/operators/SparseFiniteDifferenceOperator.py]. In TGV for example, I used both the `forward` and `backward` directions, see [here](https://github.com/epapoutsellis/CIL_and_CVXPY/blob/main/regularisers.py).

Reply:
@gfardell can this be resolved or is anything still needed?